SU2_CFD/include/numerics/CNumerics.hpp

```diff
@@ -0,0 +1,1302 @@
+  Neighbor_j;  /*!< \brief Number of neighbors of the point j. */
+  su2double
+  *Normal,       /*!< \brief Normal vector, its norm is the area of the face. */
+  *UnitNormal,   /*!< \brief Unitary normal vector. */
+  *UnitNormald;  /*!< \brief Derivative of unitary normal vector. */
+  su2double
+  TimeStep,    /*!< \brief Time step useful in dual time method. */
+  Area,        /*!< \brief Area of the face i-j. */
+  Volume;      /*!< \brief Volume of the control volume around point i. */
+  su2double vel2_inf;     /*!< \brief value of the square of freestream speed. */
+  su2double
+  *WindGust_i,  /*!< \brief Wind gust at point i. */
+  *WindGust_j;  /*!< \brief Wind gust at point j. */
+  su2double
+  *WindGustDer_i,  /*!< \brief Wind gust derivatives at point i. */
+  *WindGustDer_j;  /*!< \brief Wind gust derivatives at point j. */
+  su2double *Vorticity_i, *Vorticity_j;    /*!< \brief Vorticity. */
+  su2double StrainMag_i, StrainMag_j;      /*!< \brief Strain rate magnitude. */
+  su2double Dissipation_i, Dissipation_j;  /*!< \brief Dissipation. */
+  su2double Dissipation_ij;
+
+  su2double *l, *m;
+
+  su2double **MeanReynoldsStress; /*!< \brief Mean Reynolds stress tensor  */
+  su2double **MeanPerturbedRSM;   /*!< \brief Perturbed Reynolds stress tensor  */
+  bool using_uq;                  /*!< \brief Flag for UQ methodology  */
+  su2double PerturbedStrainMag;   /*!< \brief Strain magnitude calculated using perturbed stress tensor  */
+  unsigned short Eig_Val_Comp;    /*!< \brief Component towards which perturbation is perfromed */
+  su2double uq_delta_b;           /*!< \brief Magnitude of perturbation */
+  su2double uq_urlx;              /*!< \brief Under-relaxation factor for numerical stability */
+  bool uq_permute;                /*!< \brief Flag for eigenvector permutation */
+
+  /* Supporting data structures for the eigenspace perturbation for UQ methodology */
+  su2double **A_ij, **newA_ij, **Eig_Vec, **New_Eig_Vec, **Corners;
+  su2double *Eig_Val, *Barycentric_Coord, *New_Coord;
+
+public:
+  /*!
+   * \brief Return type used in some "ComputeResidual" overloads to give a
+   * const-view of the internally stored flux vector and Jacobians to the outside.
+   * \note The default template params make this an "all const" object, it cannot
+   * change after instantiation, nor can it be used to change the data.
+   */
+  template<class Vector_t = const su2double*,
+           class Matrix_t = const Vector_t*>
+  struct ResidualType {
+    const Vector_t residual;
+    const Matrix_t jacobian_i;
+    const Matrix_t jacobian_j;
+
+    ResidualType() = delete;
+
+    ResidualType(const Vector_t& res, const Matrix_t& jac_i, const Matrix_t& jac_j) :
+      residual(res), jacobian_i(jac_i), jacobian_j(jac_j) { }
+
+    /*!
+     * \brief The object can be directly cast to the vector type, this
+     * allows discarding the Jacobians when they are not needed.
+     */
+    operator Vector_t() { return residual; }
+  };
```

Comment:
For thread safety we need one numerics instance per thread, likewise in the call to `CNumerics::ComputeResidual` we would need dedicated copies of the residual and its Jacobians, rather than allocating multiple of those in CSolver I am allocating them in CNumerics (since we already have one per thread) and then returning this lightweight struct that gives read-only access to those members of CNumerics.